### PR TITLE
Hvg seurat v3 numba kernel

### LINKED ISSUE #3011: Update Preprocessing functions with numba
### What kind of feature would you like to request?

Additional function parameters / changed functionality / changed defaults?

### Please describe your wishes

It would be useful to switch from array operations to `numba` kernels in for most of the preprocessing. I would start writing those.

### PR DESCRIPTION
**Pull request by Intron7**
Adds a numba kernel for `seurat_v3` for sparse matrices. This kernel is a lot faster and more memory efficient. I doesn't copy nor promotes the matrix to 64-bit floats.

### CODE REVIEW

File: scanpy/preprocessing/_highly_variable_genes.py [hunk 1]
```diff
@@ -191,6 +199,37 @@ def _highly_variable_genes_seurat_v3(
...
+    batch_counts_sum = np.zeros(n_cols, dtype=np.float64)
+    for i in numba.prange(n_threads):
+        for row in range(i, len(indptr) - 1, n_threads):
+            for j in range(indptr[row], indptr[row + 1]):
+                idx = indices[j]
+                if idx >= n_cols:
+                    continue
+                element = min(np.float64(data[j]), clip_val[idx])
+                squared_sum_buffer[i, idx] += element**2
+                sum_buffer[i, idx] += element
```
Comment (flying-sheep) [inline scanpy/preprocessing/_highly_variable_genes.py]: This part looks almost identical to `sparse_mean_var_minor_axis` (the only difference seems to be the clipping)

Is there a way to deduplicate this?

File: scanpy/preprocessing/_highly_variable_genes.py [hunk 2]
```diff
@@ -191,6 +199,37 @@ def _highly_variable_genes_seurat_v3(
...
 
+@numba.njit(cache=True, parallel=True)
+def _clip_sparse(
+    indptr: NDArray[np.integer],
+    indices: NDArray[np.integer],
+    data: NDArray[np.floating],
+    *,
+    n_rows: int,
+    n_cols: int,
+    nnz: int,
```
Comment (flying-sheep) [inline scanpy/preprocessing/_highly_variable_genes.py]: n_rows and nnz are unused, should they be removed or is that a mistake?

Comment (Intron7) [reply scanpy/preprocessing/_highly_variable_genes.py]: This is not 100%final. I'm still unsure if singlethreaded might be better than multithreaded

Comment (flying-sheep) [reply scanpy/preprocessing/_highly_variable_genes.py]: OK, why would that be? Memory? Locality? I’m just guessing here haha

**Comment by Intron7** [comment]
~500k Cells.
New Version:
peak memory: 16117.61 MiB, increment: 7709.86 MiB
Wall time: 11 s
Old Version:
peak memory: 40093.42 MiB, increment: 31646.50 MiB
Wall time: 30.9 s

**Comment by eroell** [comment]
Indeed, ~3-7x faster for me & of course quite a bit more memory efficient (quickly checked with scalene).

Tests keep working (for `seurat_v3`/`seurat_v3_paper` somewhat tight numeric comparison with Seurat results), nice.

File: scanpy/preprocessing/_highly_variable_genes.py [hunk 3]
```diff
@@ -193,6 +197,26 @@ def _highly_variable_genes_seurat_v3(
         return df
 
 
+@numba.njit(cache=True)
+def _clip_sparse(
```
Comment (eroell) [inline scanpy/preprocessing/_highly_variable_genes.py]: Minor point, is the name `_clip_sparse` final..? not sure what else to call it tho, fast clip sum square anything... feel free to ignore this point if its just me, is concise enough to be clear anyways

**Review by eroell — APPROVED** [review]

File: scanpy/preprocessing/_highly_variable_genes.py (hunk 3 above)
Comment (ivirshup) [reply scanpy/preprocessing/_highly_variable_genes.py]:
```suggestion
def _sum_and_sum_squares_clipped(
```

?

File: docs/release-notes/1.10.2.md [hunk 4]
```diff
@@ -21,4 +21,5 @@
 ```
 
 * `sparse_mean_variance_axis` now uses all cores for the calculations {pr}`3015` {smaller}`S Dicks`
+* `pp.highly_variable_genes` with `flavor=seurat_v3` now uses all cores for clipping {pr}`3017` {smaller}`S Dicks`
```
Comment (ivirshup) [inline docs/release-notes/1.10.2.md]: This isn't parallel, right?

**Review by ivirshup — COMMENTED** [review]
Looks good. Kinda suprised how much faster this is.

You could probably get another big speed up by operating on the whole matrix at once, and not creating batch specific matrices.

But I think fine as is.

I added a couple points to address, but that's just naming and docs.

**Comment by ivirshup** [comment]
Would also be nice to have a timing benchmark for this.

File: docs/release-notes/1.10.2.md (hunk 4 above)
Comment (Intron7) [reply docs/release-notes/1.10.2.md]: No it's not I'll update the release note

**Comment by Intron7** [comment]
It is the whole matrix for each batch. It's just called batch_counts because I needed to make sure the format is csr

**Review by ivirshup — APPROVED** [review]
I was thinking more of getting rid of the whole `for b in np.unique(batch_info)` loop and doing the whole thing in two passes over the matrix. 

But LGTM, and I think ready to merge once those tests get fixed

**Comment by Intron7** [comment]
I don't know what makes the tests fail with dask in utils

**Pull request merged**